### PR TITLE
[occm] update listener on Octavia annotation changes

### LINKED ISSUE #1149: [occm] loadbalancer not updating when Octavia specific service annotations change
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**What happened**:
A change to any of the service annotations
* `loadbalancer.openstack.org/timeout-client-data`
* `loadbalancer.openstack.org/timeout-member-connect`
* `loadbalancer.openstack.org/timeout-member-data`
* `loadbalancer.openstack.org/timeout-tcp-inspect`
* `loadbalancer.openstack.org/x-forwarded-for`

is not picked up by the openstack-cloud-controller-manager. Changes are only propagated to the OpenStack loadbalancer listener when it is initially created.

**What you expected to happen**:
Changing one of the above service annotations should be supported and the listener updated accordingly.

**How to reproduce it**:
Add the service annotation to an existing service of type `LoadBalancer`.
```yaml
apiVersion: v1
kind: Service
metadata:
   name: test-service
   annotations:
     loadbalancer.openstack.org/timeout-client-data: "100"
     loadbalancer.openstack.org/timeout-member-connect: "100"
     loadbalancer.openstack.org/timeout-member-data: "100"
     loadbalancer.openstack.org/timeout-tcp-inspect: "100"
     loadbalancer.openstack.org/x-forwarded-for: "true"
spec:
   ports:
   - port: 22000
     protocol: TCP
     targetPort: 8080
   type: LoadBalancer
```

**Anything else we need to know?**:
The same applies for the annotation `loadbalancer.openstack.org/proxy-protocol` but updating the loadbalancer pool protocol is not supported by Octavia.
See the Octavia API pool documentation [Update a pool detail](https://docs.openstack.org/api-ref/load-balancer/v2/?expanded=update-a-pool-detail#update-a-pool).
For updating the protocol, the pool has to be deleted and recreated keeping the following in mind:
* keep the floating ip
* keep the loadbalancer and listener (if possible)
* re-create the pool (and all others we may need to recreate)

**Environment**:
- openstack-cloud-controller-manager version: master
- Openstack version: Octavia

<sub>Sean Schneeweiss <sean.schneeweiss@daimler.com>, Daimler TSS GmbH, [legal info/Impressum](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md)</sub>

### PR DESCRIPTION
**Pull request by seanschneeweiss**
**What this PR does / why we need it**:
This implements the update of the listener if one of the service annotations
* `loadbalancer.openstack.org/timeout-client-data`
* `loadbalancer.openstack.org/timeout-member-connect`
* `loadbalancer.openstack.org/timeout-member-data`
* `loadbalancer.openstack.org/timeout-tcp-inspect`
* `loadbalancer.openstack.org/x-forwarded-for`

changes.
Currently, changing one of the service annotation has no effect.

**Which issue this PR fixes(if applicable)**:
fixes #1149 

**Special notes for reviewers**:
Updating `loadbalancer.openstack.org/proxy-protocol` is currently not handled. See the issue for more details.

**Release note**:

```release-note
[openstack-cloud-controller-manager] Update listener if service annotation `loadbalancer.openstack.org/timeout-client-data`, `loadbalancer.openstack.org/timeout-member-connect`, `loadbalancer.openstack.org/timeout-member-data`, `loadbalancer.openstack.org/timeout-tcp-inspect` or `loadbalancer.openstack.org/x-forwarded-for` changes.
```

<sub>Sean Schneeweiss <sean.schneeweiss@daimler.com>, Daimler TSS GmbH, [legal info/Impressum](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md)</sub>

### CODE REVIEW

**Comment by seanschneeweiss** [comment]
@ramineni Could you kindly provide us feedback if anything is missing to approve this PR?

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 1]
```diff
@@ -1151,8 +1147,34 @@ func (lbaas *LbaasV2) ensureOctaviaListener(lbID string, oldListeners []listener
...
 		updateOpts := listeners.UpdateOpts{}
 
-		if connLimit != listener.ConnLimit {
-			updateOpts.ConnLimit = &connLimit
+		if svcConf.connLimit != listener.ConnLimit {
+			updateOpts.ConnLimit = &svcConf.connLimit
+			listenerChanged = true
+		}
+		updateOpts.InsertHeaders = &listener.InsertHeaders
+		listenerKeepClientIP := listener.InsertHeaders[annotationXForwardedFor] == "true"
```
Comment (lingxiankong) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:1155]: What if `annotationXForwardedFor` key doesn't exist?

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 2]
```diff
@@ -1297,6 +1321,11 @@ func (lbaas *LbaasV2) checkService(service *corev1.Service, nodes []*corev1.Node
 	svcConf.keepClientIP = keepClientIP
 	svcConf.enableProxyProtocol = useProxyProtocol
 
+	svcConf.timeoutClientData = getIntFromServiceAnnotation(service, ServiceAnnotationLoadBalancerTimeoutClientData, 0)
```
Comment (lingxiankong) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:1324]: Where is the default value 0 coming from? Please check the default values from Octavia API doc.

Comment (sbueringer) [reply pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:1324]: Isn't it just exactly the same as before? In the old code l.1126 the annotations were read if set. If they were not set the go default (which is 0) was used?

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go (hunk 1 above)
Comment (sbueringer) [reply pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:1155]: Then `listenerKeepClientIP` is false, which should be the same as when then key is set to "false"

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go (hunk 2 above)
Comment (seanschneeweiss) [reply pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:1324]: Thank you very much for this hint. From the [Octavia API docs](https://docs.openstack.org/api-ref/load-balancer/v2/?expanded=create-listener-detail#create-listener).

Name | In | Type | Description
-- | -- | -- | --
timeout_client_data (Optional) | body | integer | Frontend client inactivity timeout in milliseconds. Default: **50000**. New in version 2.1
timeout_member_connect (Optional) | body | integer | Backend member connection timeout in milliseconds. Default: **5000**. New in version 2.1
timeout_member_data (Optional) | body | integer | Backend member inactivity timeout in milliseconds. Default: **50000**. New in version 2.1
timeout_tcp_inspect (Optional) | body | integer | Time, in milliseconds, to wait for additional TCP packets for content inspection. Default: **0**. New in version 2.1

Will change the defaults accordingly.

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go (hunk 1 above)
Comment (seanschneeweiss) [reply pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:1155]: I agree with @sbueringer: If `annotationXForwardedFor` does not exist, `listenerKeepClientIP` will be `false`.

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 3]
```diff
@@ -592,19 +599,24 @@ func getStringFromServiceAnnotation(service *corev1.Service, annotationKey strin
...
-func getIntFromServiceAnnotation(service *corev1.Service, annotationKey string) (int, bool) {
-	intString := getStringFromServiceAnnotation(service, annotationKey, "")
-	if len(intString) > 0 {
-		annotationValue, err := strconv.Atoi(intString)
-		if err == nil {
-			klog.V(4).Infof("Found a Service Annotation: %v = %v", annotationKey, annotationValue)
-			return annotationValue, true
+//getIntFromServiceAnnotation searches a given v1.Service for a specific annotationKey and either returns the annotation's integer value or a specified defaultSetting
+func getIntFromServiceAnnotation(service *corev1.Service, annotationKey string, defaultSetting int) int {
+	klog.V(4).Infof("getIntFromServiceAnnotation(%v, %v, %v)", service, annotationKey, defaultSetting)
```
[the remark this answers is not in the document]
Comment (seanschneeweiss) [reply pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:604]: The log was kept analogously to `getBoolFromServiceAnnotation`. But I agree to remove this line.

**Comment by seanschneeweiss** [comment]
@ramineni Unfortunately the jobs failed due to `git clone https://github.com/kubernetes/kubernetes /home/zuul/src/k8s.io/kubernetes -b master` which requires go 1.15.0 since this [commit](https://github.com/kubernetes/kubernetes/commit/bdcffdced38d0e692e81e7dbc0df444639e4ca5b#diff-500746274e2e9471f21c105f5d0a91a6) on Aug. 13.
```
Detected go version: go version go1.14.4 linux/amd64.
Kubernetes requires go1.15.0 or greater.
Please install go1.15.0 or later.
```

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go (hunk 2 above)
Comment (sbueringer) [reply pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Ah I think I got it wrong. In the old version the values were only propagated to Openstack when the annotations are set. Now we set them in any case so it makes sense to set the defaults

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go (hunk 1 above)
Comment (lingxiankong) [reply pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:1154]: Yes, I was thinking about Python...

**Review by lingxiankong — APPROVED** [review]
/lgtm

**Comment by lingxiankong** [comment]
need rebase

**Pull request merged**